mm2src/coins/nft/storage/sql_storage.rs

```diff
@@ -369,37 +548,75 @@ fn block_number_from_row(row: &Row<'_>) -> Result<i64, SqlError> { row.get::<_,
 
 fn nft_amount_from_row(row: &Row<'_>) -> Result<String, SqlError> { row.get(0) }
 
-fn get_transfers_from_block_builder<'a>(
+fn get_nfts_by_token_address_statement<'a, F>(
     conn: &'a Connection,
     chain: &'a Chain,
-    from_block: u64,
-) -> MmResult<SqlQuery<'a>, SqlError> {
-    let table_name = nft_transfer_history_table_name(chain);
+    table_name_creator: F,
+) -> MmResult<Statement<'a>, SqlError>
+where
+    F: FnOnce(&Chain) -> String,
+{
+    let table_name = table_name_creator(chain);
     validate_table_name(table_name.as_str())?;
-    let mut sql_builder = SqlQuery::select_from(conn, table_name.as_str())?;
-    sql_builder
-        .sql_builder()
-        .and_where(format!("block_number >= '{}'", from_block))
-        .order_asc("block_number")
-        .field("details_json");
-    drop_mutability!(sql_builder);
-    Ok(sql_builder)
+    let sql_query = format!("SELECT * FROM {} WHERE token_address = ?", table_name);
+    let stmt = conn.prepare(&sql_query)?;
+    Ok(stmt)
 }
 
-fn get_transfers_by_token_addr_id_statement<'a>(
+fn get_token_addresses_statement<'a, F>(
     conn: &'a Connection,
     chain: &'a Chain,
-) -> MmResult<Statement<'a>, SqlError> {
+    table_name_creator: F,
+) -> MmResult<Statement<'a>, SqlError>
+where
+    F: FnOnce(&Chain) -> String,
+{
+    let table_name = table_name_creator(chain);
+    validate_table_name(table_name.as_str())?;
+    let sql_query = format!("SELECT DISTINCT token_address FROM {}", table_name);
+    let stmt = conn.prepare(&sql_query)?;
+    Ok(stmt)
+}
+
+// fn get_transfers_from_block_builder<'a>(
+//     conn: &'a Connection,
+//     chain: &'a Chain,
+//     from_block: u64,
+// ) -> MmResult<SqlQuery<'a>, SqlError> {
+//     let table_name = nft_transfer_history_table_name(chain);
+//     validate_table_name(table_name.as_str())?;
+//     let mut sql_builder = SqlQuery::select_from(conn, table_name.as_str())?;
+//     sql_builder
+//         .sql_builder()
+//         .and_where(format!("block_number >= '{}'", from_block))
+//         .order_asc("block_number")
+//         .field("*");
+//     drop_mutability!(sql_builder);
+//     Ok(sql_builder)
+// }
+
+fn get_transfers_from_block_statement<'a>(conn: &'a Connection, chain: &'a Chain) -> MmResult<Statement<'a>, SqlError> {
     let table_name = nft_transfer_history_table_name(chain);
     validate_table_name(table_name.as_str())?;
     let sql_query = format!(
-        "SELECT details_json FROM {} WHERE token_address = ? AND token_id = ?",
+        "SELECT * FROM {} WHERE block_number >= ? ORDER BY block_number ASC",
         table_name
     );
     let stmt = conn.prepare(&sql_query)?;
     Ok(stmt)
 }
 
+fn get_transfers_by_token_addr_id_statement<'a>(
+    conn: &'a Connection,
+    chain: &'a Chain,
+) -> MmResult<Statement<'a>, SqlError> {
+    let table_name = nft_transfer_history_table_name(chain);
+    validate_table_name(table_name.as_str())?;
```

Comment:
This is duplicated a lot, and I am not even sure if we ensure validating the table name everytime we generate them with `nft_transfer_history_table_name(chain: &Chain)`. But if you implement something like `to_table_name -> SqlResult<String> or Option<String>` to `Chain`, we can do `let table_name = chain.to_table_name()?` and do the validation inside of that function so we don't need to worry about if we missed validations or not.

Reply:
Thanks for the note, done